File: python/behavioral/strategy.py
#!/usr/bin/python
# -*- coding : utf-8 -*-

"""
brief:
  Strategy. 
    Define a family of algorithms, encapsulate each 
  one, and make them interchangeable.
  
         Context      ->              Strategy
                                      |     |
   ConcreteContext     ConcreteStrategy1  ConcreteStrategy2
"""

import types

# Context
class Context:
  def __init__(self, func=None):
    if func is not None:
      self.execute = types.MethodType(func, self)

  def execute(self):
    print("No strategy")

# ConcreteStrategy1
def strategy1(self):
  print("Strategy1")

# ConcreteStrategy2
def strategy2(self):
  print("Strategy2")

if __name__ == '__main__':
  s0 = Context()
  s1 = Context(strategy1)
  s2 = Context(strategy2)

  s0.execute()
  s1.execute()
  s2.execute()

File: design-patterns/cpp/behavioral/strategy.cpp
/*!
* \brief   Strategy. Define a family of algorithms, encapsulate each 
*       one, and make them interchangeable.
*
*          Context      ->              Strategy
*                                       |     |
*     ConcreteContext     ConcreteStrategy1  ConcreteStrategy2
*/

#include <iostream>

class Strategy {
public:
  virtual std::string Execute() = 0;
};

// ConcreteStrategy1.
class Walk : public Strategy {
public:
  std::string Execute() {
    return "<Walk>";
  }
};

// ConcreteStrategy2.
class Swim :public Strategy {
public:
  std::string Execute() {
    return "<Swim>";
  }
};

// Context.
class Context {
public: 
  virtual void SetStrategy(Strategy *w) = 0;
  virtual void Go() = 0;
};

// ConcreteContext.
class Go2Beijing : public Context {
public:
  Go2Beijing() : way_(0) {};
  void SetStrategy(Strategy *w) {
    way_ = w;
  }
  void Go() {
    if(way_ != 0)
      std::cout << "I'm going to " << way_->Execute().c_str() << " to Beijing." << std::endl;
    else
      std::cout << "I don't know how to get to Beijing yet." << std::endl;
  }

private:
  Strategy *way_;
};

int main() {
  Strategy *swim = new Swim();
  Strategy *walk = new Walk();

  Context *context = new Go2Beijing();

  context->Go();
  std::cout << "/////////////////" << std::endl;

  context->SetStrategy(swim);
  context->Go();
  std::cout << "/////////////////" << std::endl;

  context->SetStrategy(walk);
  context->Go();
  std::cout << "/////////////////" << std::endl;

  return 0;
}
